feat(crank): add describe command

jbasement opened the pull request:
# Description of your changes

This PR adds the `describe` command to the crank/ crossplane CLI. The command is powered by the `graph` package.

## Commands 
`describe` - Describes a given resource (usually a claim or composite) by getting the manifest of the resource and all its children. The resource is then printed out by either a CLI table or a graphviz png. The user has the option to set the printed fields, output type and a few more parameters.

## Packages
`graph` - The graph package is able to get a resource (usually a claim or composite resource) from k8s, including all its children. The graph package implements a printer interface and currently comes with 3 printers to return the resource. A tree (default), table, or dot (must be rendered) printer.
This builds the foundation to implement more advanced commands like e.g. `diagnose`. Additional steps can be implemented between getting a resource and printing it out, like a diagnosis function.

Partially addresses #4672, with the `diagnose` command to come in a later PR

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Added or updated unit **and** E2E tests for my change.
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] ~Added `backport release-x.y` labels to auto-backport this PR, if necessary.~
- [ ] ~Opened a PR updating the [docs], if necessary.~

## Note for reviewer
- I didnt implement unit tests for `resource.go` yet as I'm not sure yet if and how much rework it still needs
- I didnt implement the logger to the packages yet as I wasnt sure what is the best way to implement it there. It would be happy to get a suggestion there how I could implement it

# Screenshots
## `describe` command with different output formats

CLI tree
![image](https://github.com/crossplane/crossplane/assets/40320614/b776fd76-3992-49ec-819e-50d40fb39eec)


CLI table
![image](https://github.com/crossplane/crossplane/assets/40320614/4f5fd81b-ff11-4ad1-9099-81f32b9b723d)


GraphViz png
`crank describe objectstorage test-bucket-00002 -o graph | dot -Tpng  > test.png && open test.png`
![image](https://github.com/crossplane/crossplane/assets/40320614/c4ffd7b0-0ca7-4028-a435-5892d1778bd9)

phisco commented:
Hi @jbasement, thanks for your contribution!
Could you add to the description screenshots of the describe and diagnose commands outputs, that would help people having a better grasp of what you are proposing without having to actually run it first. 🙏 
Then, would it make sense to address the two subcommands one at a time in two dedicated PRs, just to make the process smoother?

jbasement commented:
Good idea, I added some screenshots to the PR description.
Sure I could split up the PR for the two commands, but both commands depend on the packages added in this PR. So if we split up the PRs first this PR has to be merged before we can implement the second command. If we still want to go ahead and split the PR in two, I'd suggest first implementing the `describe` command as it gives a better overview of the k8s resource and then in a second PR implement the `diagnose` command. wdyt?

phisco commented:
Yes, @jbasement, I think splitting, even if they have to be serialized, would make the review process easier. Let's first get a good describe command and then on top of that add the diagnose functionalities 😉

jbasement commented:
@phisco Agreed that makes sense! :) Updated the branch and the PR description accordingly. Keeping the branch with the diagnose command local for now until this PR is merged.

jbw976 commented on internal/printer/graphviz.go:
```diff
@@ -0,0 +1,98 @@
+
+	"github.com/crossplane/crossplane/internal/k8s"
+	"github.com/emicklei/dot"
+	"github.com/goccy/go-graphviz"
```
i wanted to take this PR for a test spin to get a better feel for the functionality, but I'm having trouble building on my Mac M1 Max:

```
❯ make
<omitted...>
22:33:30 [ .. ] go build darwin_arm64
github.com/crossplane/crossplane/cmd/crossplane
package github.com/crossplane/crossplane/cmd/crank
	imports github.com/crossplane/crossplane/internal/printer
	imports github.com/goccy/go-graphviz
	imports github.com/goccy/go-graphviz/cgraph
	imports github.com/goccy/go-graphviz/internal/ccall: build constraints exclude all Go files in /Users/jared/go/pkg/mod/github.com/goccy/go-graphviz@v0.1.1/internal/ccall
22:33:33 [FAIL]
```

Is there additional build or environment config needed to build this? I'm not sure if we're already using cgo, but I'm  hesitant to add that complexity to our build and dependencies if we don't absolutely have to. e.g. https://relistan.com/cgo-when-and-when-not-to-use-it

jbasement replied:
Hm no I didn't set any specific build or environment config for go-graphviz to work. 
But coming back to your point about not using cgo, we can make it work without the go-graphviz library. The graph file is built using the dot library. This library doesn't support saving the graph as a PNG that's why I started using go-graphviz in the first place. But the graph can be rendered like this by the user.`<my-command> | dot -Tpng  > test.png && open test.png` 
https://github.com/emicklei/dot#display-your-graph

jbw976 replied:
+1 for offloading the rendering / image generation to an external tool. @phisco mentioned that this is symmetrical with what the golang tools do as well and it would remove complexity from our build/dependencies.

phisco replied:
Referring to https://go.dev/ref/mod#go-mod-graph

jbasement replied:
Also thinking about it now, it is easier to test if the function is working by comparing the dot output instead of checking a PNG file. Reworked the function now as discussed and removed the graphviz library

jbw976 commented on internal/printer/cli.go:
```diff
@@ -0,0 +1,76 @@
+)
+
+// Prints out a CLI table of the passed Resource. The fields variable determines the header and values of the table.
+func CliTable(rootResource k8s.Resource, fields []string) error {
```
I'm having trouble personally parsing the table view as that format doesn't effectively visually capture the relationships between the resources in the table.

Is it possible to instead pursue a tree view? That would very clearly demonstrate the hierarchy and relationships to the user and overall be more effective. That may be much harder to implement though...what do you think?

phisco replied:
+1

jbasement replied:
Sure we can try to implement a tree view instead of/ additionally to the table. What do you have in mind for the tree view? Something like kube-lineage, which is a mix of table and tree? Or really only a plain [tree](https://github.com/xlab/treeprint) that instead of file and dir names has the resource information?

jbw976 replied:
As discussed in the SIG-DevEx meeting today, a [kube-lineage](https://github.com/tohjustin/kube-lineage) hybrid tree/table view would be ideal, but is probably too difficult to implement before feature freeze.  Let's start with a very simple tree view that shows kind/name with hierarchical indentation and we can always make it more fancy later.

Let's also keep the table view too, no need to throw that work away, but let's make the tree view as the default experience.

jbasement replied:
Added the tree output to the printer package and set it as default. I added a screenshot of how it looks atm. The user also has the option to set the fields as in the other output methods. Please let me know how you like the current design of the output. @phisco @jbw976 

I'll give the kube-lineage-like tree/table a try this weekend but will focus primarily on any feedback for the existing implementation in this PR.

jbw976 reviewed (commented):
very very **very** cool to see your initiative @jbasement and your willingness to start contributing to improving the developer experience in Crossplane! Big kudos for jumping in here and sharing your efforts with the community! 🙇‍♂️ 

I love the idea here but I do have a couple high level questions about the experience that we'll want to figure out first, see comments for details.

* table view vs. tree view
* usage of cgo

Really looking forward to your contributions going forward too 😉

jbw976 reviewed (commented):
This is really good @jbasement, I am loving the tree view approach that we default to now! Thanks for taking the directional feedback and incorporating that into the experience and thanks again for your efforts on this. A true member of SIG-DevEx! 🤩 😂  

There are multiple failing checks for the PR status, let's see if we can get those all fixed.  A great start to test those locally will be running `make reviewable`.

I have a few comments to consider, but I would actually love to get this merged pretty quickly and not spend a ton of time on a longer tail of work in this PR. This is because I think the `crossplane diagnose` command is even more important than this, and getting that in before feature freeze on Wednesday would be the best possible scenario in my book 🤓 

@phisco if you could also help get this PR to an approved state so we can merge soon and focus on `crossplane diagnose`, that would be awesome! 🙇‍♂️

jbasement commented on internal/printer/cliTable_test.go:
```diff
@@ -0,0 +1,117 @@
+package printer
+
+// import (
```
I commented them out as I was implementing the Printer interface and didn't adjust the tests accordingly yet. But going to fix the tests today

jbasement commented on internal/printer/graph_test.go:
```diff
@@ -0,0 +1,84 @@
+package printer
+
+// import (
```
Yes going to fix them today

jbasement commented on internal/printer/cliTable.go:
```diff
@@ -0,0 +1,92 @@
+package printer
```
Agreed, I moved the printer to `cmd/crank/printer`

jbasement commented on internal/k8s/resource.go:
```diff
@@ -0,0 +1,347 @@
+}
+
+type Resource struct {
+	Manifest           *unstructured.Unstructured
```
Yes you're right same for `LatestEventMessage`, I also set it to private. I left `Children` exported for now as they have to be accessed in the printer functions, but we could implement helper functions for the Children in the future as well

jbasement replied:
I export the vars of the struct so I can init a complete struct for the tests in the `printer` package. In the future, I could maybe add a `CreateDummyResource` function that returns Resources for testing.
But let me know if there is a better or simpler way to solve this :)

jbasement replied:
Merged the packages and made manifest a local now

jbasement commented on internal/k8s/resource.go:
```diff
@@ -0,0 +1,347 @@
+	conditions, _, _ := unstructured.NestedSlice(r.Manifest.Object, "status", "conditions")
+	for _, condition := range conditions {
+		conditionMap, _ := condition.(map[string]interface{})
+		conditionType, _ := conditionMap["type"].(string)
```
You're right I added a check to all assertions.
`type` should be fine ([source](https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L1507)), still added the check in case the `type` field might change in the future

jbasement commented on internal/k8s/resource.go:
```diff
@@ -0,0 +1,347 @@
+// The main function of resource. Returns a Resource and all its child resources.
+func GetResource(resourceKind string, resourceName string, namespace string, kubeconfig *rest.Config) (*Resource, error) {
+	// Init new client
+	client, err := newClient(kubeconfig)
```
Good point. Moved the client init away from the k8s package

phisco commented on cmd/crank/describe.go:
```diff
@@ -0,0 +1,96 @@
+		return errors.Wrap(err, errGetResource)
+	}
+
+	// Configure printer
+	var p printer.Printer
+
+	switch c.Output {
+	case "tree":
+		p = &printer.TreePrinter{
+			Indent: "",
+			IsLast: true,
+		}
+	case "table":
+		p = &printer.TablePrinter{}
+	case "graph":
+		p = &printer.GraphPrinter{}
+	}
```
I'd probably make "tree", "table" and "graph" constants with a dedicated type and move this switch to the printer package, doing also there the validation of the known types, something like:
```golang
func New(type string) (Printer, error) {
  switch Type(type) {
    case TypeTree:
      return &printer.TreePrinter{
			Indent: "",
			IsLast: true,
		}, nil
... etc ...
  default:
    return nil, errors.Errorf("unknown printer output type: %s", type)
  }
  ...
}
```

jbasement replied:
I implemented your suggestion in this commit https://github.com/crossplane/crossplane/pull/4751/commits/30b979234e2b9b2dcacab017f82784ec4d25d952. Let me know if it is implemented as expected.

jbasement replied:
And small fix https://github.com/crossplane/crossplane/pull/4751/commits/bf575090521f5d82f7829d355fb6f08394fbb547

phisco replied:
Yes, that's exactly what I meant! Looks good 🙏

phisco commented on cmd/crank/describe.go:
```diff
@@ -0,0 +1,96 @@
+func (c *describeCmd) Run(logger logging.Logger) error {
+	logger = logger.WithValues("Kind", c.Kind, "Name", c.Name)
+
+	allowedFields := []string{"parent", "name", "kind", "namespace", "apiversion", "synced", "ready", "message", "event"}
```
I would suggest moving this to be a global variable

jbasement replied:
Done

phisco commented on cmd/crank/describe.go:
```diff
@@ -0,0 +1,96 @@
+	logger = logger.WithValues("Kind", c.Kind, "Name", c.Name)
+
+	allowedFields := []string{"parent", "name", "kind", "namespace", "apiversion", "synced", "ready", "message", "event"}
+	allowedOutput := []string{"tree", "table", "graph"}
```
see comment below about the allowed values and validation

jbasement replied:
Commented above

phisco reviewed (commented):
left a few comments 👍

jbasement commented on internal/k8s/resource.go:
Yes, I agree that's a really good idea. Implemented the changes as suggested

jbasement commented on internal/k8s/resource.go:
```diff
@@ -0,0 +1,362 @@
+// GetAPIVersion returns resource apiversion as string
+func (r *Resource) GetAPIVersion() string {
+	return r.Manifest.GetAPIVersion()
+}
```
Good point I should've checked that 👍

jbasement commented:
@phisco Thanks for the thorough review! Regarding the 3 open comments, I don't think I'll find the time before the code freeze to implement these. Not sure if they are showstoppers or if someone else might be able to implement those in time?

jbw976 commented:
@phisco is there anything remaining that must be fixed before we can approve/merge this PR? It would be great to get it in and then iterate to add the `diagnose` functionality, which IMO is the most important feature in this particular area 🤩

jbasement commented:
> I think something broke as it's not able to lookup a claim: <img alt="image" width="1279" src="https://user-images.githubusercontent.com/5697904/275770628-37312da3-4ebb-4dcc-9264-bf3b441ead1e.png">
> 
> To reproduce:
> 
> * just install the latest helm chart (v1.13.2): `helm upgrade --install crossplane crossplane-stable/crossplane --namespace crossplane-system --create-namespace --set args='{--debug}'`
> * apply the manifests in the manifests from the e2es:`kubectl apply -f test/e2e/manifests/apiextensions/composition/minimal/setup`
> * create the claim in the default namespace: `kubectl apply -f test/e2e/manifests/apiextensions/composition/minimal/claim.yaml`

I think the issue is a typo in your command when executing. You are referencing `....example.or` instead of `example.org`. But would be a nice feature to have a suggestion like kubectl has

phisco commented:
good call 😅 however I think that was just the latest attempt and it was already failing given that it's still failing now:
<img width="3007" alt="image" src="https://github.com/crossplane/crossplane/assets/5697904/ed793883-765c-41ea-be3d-b4dc2467f5ff">

Works fine for composite resources, but not for claims, I guess some issue with the namespace? 🤔

jbasement commented:
@phisco I was able to fix the issues.
The issue was related to the `isResourceNamespaced` function and the other issue related to the restmapper.
It mapped the passed group `nopresources.nop.example.org` to the similar group `nopresources.nop.crossplane.io`. This surely can be solved in a better way than I did but it works for now. My change added a bit of (probably unnecessary) complexity to the function now

phisco commented:
Awesome @jbasement! works like a charm now, thank you!
